File: binary search/binarysearch.cpp
#include <stdio.h>
#include <string>
#include <iostream>

int search(int value, int array[], int length) {
	if (array == NULL) {
		return -1;
	}

	int left = 0;
	int right = length - 1;

	while (left <= right) {
		int middle = left + ((right - left) >> 1);
		if (array[middle] > value) {
			right = middle - 1;
		}
		else if (array[middle] < value) {
			left = middle + 1;
		}
		else {
			return middle;
		}
	}
	return -1;
}

template <typename T>
int genericSearch(T value, T array[], int length) {
	if (array == NULL) {
		return -1;
	}
	int left = 0;
	int right = length - 1;

	while (left <= right) {
		int middle = left + ((right - left) >> 1);
		if (array[middle] > value) {
			right = middle - 1;
		}
		else if (array[middle] < value) {
			left = middle + 1;
		}
		else {
			return middle;
		}
	}
	return -1;
}


int main() {
	int a[6] = {1, 3, 5, 7, 9, 25};
	std::cout << search(3, a, 6) << " should be equals to 1" << std::endl;

	std::string b[4] = {"something", "testing", "uva", "uzw"};
	std::string v = "uzw";
	std::cout << genericSearch(v, b, 4) << " should be equal to 3" << std::endl;

	return 0;
}

File: src/main/java/search/BinarySearch.java
package search;

public class BinarySearch {
    public static void main(String[] args) {
        int[] array = new int[]{-1, 3, 7, 9, 12, 13, 14, 16};
        System.out.println(search(array, 9));
    }

    public static int search(int[] nums, int target) {
        if (nums.length == 0) {
            return -1;
        }

        int left = 0, right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] < target) {
                left = mid + 1;
            } else if (nums[mid] > target) {
                right = mid - 1;
            } else {
                return mid;
            }
        }

        return -1;
    }
}
